ee/clickhouse/queries/property_optimizer.py

```diff
@@ -0,0 +1,112 @@
+from typing import Dict, List, Optional, Set, Tuple, Union, cast
+
+from rest_framework.exceptions import ValidationError
+
+from posthog.constants import PropertyOperatorType
+from posthog.models import Filter
+from posthog.models.entity import Entity
+from posthog.models.filters.mixins.utils import cached_property
+from posthog.models.filters.path_filter import PathFilter
+from posthog.models.filters.retention_filter import RetentionFilter
+from posthog.models.filters.stickiness_filter import StickinessFilter
+from posthog.models.property import Property, PropertyGroup
+
+
+class PropertyOptimizer:
+    """
+    This class is responsible for figuring out what person or group properties can and should be pushed down to their
+    respective tables in the query filter.
+
+    This speeds up queries since clickhouse ends up selecting less data.
+    """
+
+    def __init__(self, filter: Union[Filter, PathFilter, RetentionFilter, StickinessFilter], team_id: int):
+        self.filter = filter
+        self.team_id = team_id
+
+    @cached_property
+    def property_groups(self):
```

Comment:
Suggestion: Return a dataclass here instead of a tuple with relevant field names